images/metering-ansible-operator/roles/meteringconfig/library/k8s_event.py

```diff
@@ -0,0 +1,382 @@
+# now = datetime.datetime.now()
+# rfc = kubernetes.config.dateutil.format_rfc3339(now)
+
+EVENT_ARG_SPEC = {
+    "state": {"default": "present", "choices": ["present", "absent"],},
+    "name": {"required": True},
+    "namespace": {"required": True},
+    "merge_type": {"type": "list", "choices": ["json", "merge", "strategic-merge"]},
+    "message": {"type": "str", "required": True},
+    "reason": {"type": "str", "required": True},
+    "reportingComponent": {"type": "str", "required": True},
+    "type": {"choices": ["Normal", "Warning"],},
+    "source": {"type": "str", "component": {"type": "str", "required": True}},
+    "involvedObject": {
+        "type": "str",
+        "apiVersion": {"type": "str", "required": True},
+        "kind": {"type": "str", "required": True},
+        "name": {"type": "str", "required": True},
+        "namespace": {"type": "str", "required": True},
+    },
+}
+
+
+class KubernetesEvent(KubernetesRawModule):
+    def __init__(self, *args, **kwargs):
+        super(KubernetesEvent, self).__init__(*args, k8s_kind="Event", **kwargs)
+
+    @property
+    def argspec(self):
+        """ argspec property builder """
+        argument_spec = copy.deepcopy(AUTH_ARG_SPEC)
+        argument_spec.update(EVENT_ARG_SPEC)
+        return argument_spec
+
+    def execute_module(self):
+        """ Module execution """
+        self.client = self.get_api_client()  # mostly auth
+
+        api_version = "v1"
+        name_args = self.params.get("name")
+        namespace_args = self.params.get("namespace")
+        message = self.params.get("message")
+        reason = self.params.get("reason")
+        reporting_component = self.params.get("reportingComponent")
+        event_type = self.params.get("type")
+        source = self.params.get("source")
+
+        definition = defaultdict(defaultdict)
+
+        def_meta = definition["metadata"]
+        def_meta["name"] = self.params.get("name")
+        def_meta["namespace"] = self.params.get("namespace")
+
+        def_involvedObject = definition["involvedObject"]
+        def_involvedObject["namespace"] = self.params.get("namespace")
+        def_involvedObject["apiVersion"] = self.params.get("apiVersion")
+        def_involvedObject["kind"] = self.params.get("kind")
+        def_involvedObject["name"] = self.params.get("name")
+        def_involvedObject["uid"] = self.params.get("uid")
+        def_involvedObject["resourceVersion"] = self.params.get("resourceVersion")
+
+        resource = self.find_resource(
+            "Event", "v1", fail=True
+        )  # finds resource or gets it from the client api, looks for resrouce
+        # second call to find resource for involved object, set kind to arg of involved OBject kind
+        # rnewesource = self.find_resource('involvedObjectKind', 'v1', fail=True)
+
+        # this just establishes tha vars and seperates out the reason and count from returned cli object
+        priorCount = 1
+        try:
```

Comment:
All 2 of these `try` `except` blocks can be consolidated. You're querying the same thing multiple times unnecessarily. It would be simpler if you just did something like this:

```
priorEvent = None
# Query the API for the existing event. If it doesn't exist, priorEvent will be None.
try:
    priorEvent = resource.get(
        name=meta["name"],
        namespace=meta["namespace"]
    )
except openshift.dynamic.exceptions.NotFoundError:
    pass

# set some default values for timestamps/count for when priorEvent doesn't exist
firstTimestamp = now
lastTimestamp = now
count = 1

# handle if priorEvent exists
if priorEvent:
   # do stuff when reason changed
   if priorEvent["reason"] != reason:
       # do stuff related count when reason changed
       count = 1

       # do stuff related timestamps when reason changed
       firstTimestamp = ....
       lastTimestamp = ...
   else: # do stuff when reason doesn't change
       # do stuff related count when reason is the same
       count = priorEvent["count"] + 1
       # do stuff related timestamps when reason is the same
       firstTimestamp = ....
       lastTimestamp = ...
```